### PR TITLE
Handle `macro_rules!` tokens consistently across crates

### PR DESCRIPTION
**Pull request by Aaron1011**
When we serialize a `macro_rules!` macro, we used a 'lowered' `TokenStream` for its body, which has all `Nonterminal`s expanded in-place via `nt_to_tokenstream`. This matters when an 'outer' `macro_rules!` macro expands to an 'inner' `macro_rules!` macro - the inner macro may use tokens captured from the 'outer' macro in its definition.

This means that invoking a foreign `macro_rules!` macro may use a different body `TokenStream` than when the same `macro_rules!` macro is invoked in the same crate. This difference is observable by proc-macros invoked by a `macro_rules!` macro - a `None`-delimited group will be seen in the same-crate case (inserted when convering `Nonterminal`s to the `proc_macro` crate's structs), but no `None`-delimited group in the cross-crate case.

To fix this inconsistency, we now insert `None`-delimited groups when 'lowering' a `Nonterminal` `macro_rules!` body, just as we do in `proc_macro_server`. Additionally, we no longer print extra spaces for `None`-delimited groups - as far as pretty-printing is concerned, they don't exist (only their contents do). This ensures that `Display` output of a `TokenStream` does not depend on which crate a `macro_rules!` macro was invoked from.

This PR is necessary in order to patch the `solana-genesis-programs` for the upcoming hygiene serialization breakage (https://github.com/rust-lang/rust/pull/72121#issuecomment-646924847). The `solana-genesis-programs` crate will need to use a proc macro to re-span certain tokens in a nested `macro_rules!`, which requires us to consistently use a `None`-delimited group.

See `src/test/ui/proc-macro/nested-macro-rules.rs` for an example of the kind of nested `macro_rules!` affected by this crate.

### CODE REVIEW

**Comment by Aaron1011** [comment]
I'm not quite sure what the best way to handle pretty-printing is. There seem to be a few options:

1. Don't insert `None`-delimited groups, and leave `macro_rules!` behavior inconsistent across crates. This seems confusing for users, and has real consequences when writing proc-macros (less span information is available). However, it avoids any breakage.
2. Do what this PR currently does. but special-case `stringify!` to not insert extra spaces for outer (or maybe all?) `Nonterminal`s. This seems like kind of a hack, and will change the behavior of proc-macros that generate calls to `stringify!` with custom tokens.
3. Instead of adding spaces when pretty-printing `None`-delimited groups, just print nothing. I think this may be the best option: since they will be lost on re-parsing, they only serve as an indication that a metavariable replacement occurred (but without providing any additional information). I think it's almost impossible to write any code which actually uses the behavior, and any code which does would be better served by inspecting the `TokenStream` directly. However, this will probably result in some amount of breakage due to crates matching exactly on the pretty-printed output.

**Comment by ratijas** [comment]
Hi,

Just throwing my two cents here. I heard about `None`-delimited groups which you are talking about. In fact, I had troubles with them, and found some bugs in rustc AND proc_macro ecosystem: #71422.

Any chances that this PR fixes some of these issues raised there?

**Comment by Aaron1011** [comment]
@ratijas: I don't think this PR will affect #71422, since that issue don't seem to be related to nested `macro_rules!` macros.

**Comment by Aaron1011** [comment]
I've switched to removing the extra spaces printed for `None`-delimited groups. This has the nice effect of making re-collecting tokens a no-op in more cases - the pretty-printed output is no longer affected by the replacement of `Nonterminal`s with `None`-delimited groups. We should probably do a Crater run for this.

**Comment by petrochenkov** [comment]
>To fix this inconsistency, we now insert `None`-delimited groups when 'lowering' a `Nonterminal` `macro_rules!` body, just as we do in `proc_macro_server`.

This is a great catch, I never noticed we forget to wrap tokens into a group during "lowering" nonterminals to HIR.
Serializing them without wrapping is certainly incorrect and may cause mistakes like `$e * 3` where `$e` is `1 * 2` turning into `1 + 2 * 3`.

One problem here is that rustc parser is entirely unready to encounter `None` delimiters and will interpret them as "no parents" anyway - https://github.com/rust-lang/rust/issues/67062, it shouldn't block this fix though.

File: src/librustc_ast/attr/mod.rs [hunk 1]
```diff
@@ -590,6 +593,11 @@ impl MetaItemKind {
                 tokens.next();
                 MetaItemKind::list_from_tokens(inner_tokens)
             }
+            Some(TokenTree::Delimited(_, token::NoDelim, inner_tokens)) => {
```
Comment (petrochenkov) [inline src/librustc_ast/attr/mod.rs:596]: Have you seen this case in code? Could you add a test case?
I'm not sure how could `#[attr ⟪ 2 + 2 ⟫]` get through parser.

File: src/librustc_ast_pretty/pprust.rs [hunk 2]
```diff
@@ -245,7 +245,7 @@ fn token_kind_to_string_ext(tok: &TokenKind, convert_dollar_crate: Option<Span>)
         token::CloseDelim(token::Bracket) => "]".to_string(),
         token::OpenDelim(token::Brace) => "{".to_string(),
         token::CloseDelim(token::Brace) => "}".to_string(),
-        token::OpenDelim(token::NoDelim) | token::CloseDelim(token::NoDelim) => " ".to_string(),
+        token::OpenDelim(token::NoDelim) | token::CloseDelim(token::NoDelim) => "".to_string(),
```
Comment (petrochenkov) [inline src/librustc_ast_pretty/pprust.rs:248]: Need a test for a nodelim group between some other tokens.
I'm afraid that `A ⟪B⟫ C` can be printed as `ABC` sometimes turning three tokens into one or two.

File: src/librustc_ast/attr/mod.rs (hunk 1 above)
Comment (Aaron1011) [reply src/librustc_ast/attr/mod.rs:596]: This was needed to get libstd to compile - I'll try to minimize the case I found.

**Comment by Aaron1011** [comment]
@petrochenkov: I've added additional test cases

File: src/librustc_ast/attr/mod.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustc_ast/attr/mod.rs:596]: The added test with `#[doc = $x]` corresponds to the other change above though.

Comment (Aaron1011) [reply src/librustc_ast/attr/mod.rs]: You're right - I've removed this match arm.

**Comment by Aaron1011** [comment]
@craterbot check

**Comment by pietroalbini** [comment]
@craterbot abort

**Comment by pietroalbini** [comment]
@craterbot check

**Comment by Aaron1011** [comment]
@craterbot check

**Comment by Aaron1011** [comment]
@craterbot abort

**Comment by Aaron1011** [comment]
@craterbot check

**Comment by Aaron1011** [comment]
@craterbot abort

**Comment by pietroalbini** [comment]
@craterbot retry

**Comment by pietroalbini** [comment]
@craterbot check

**Comment by pietroalbini** [comment]
@craterbot abort name=pr-73569

**Comment by Aaron1011** [comment]
There is one legitimate regression: [dmutil](https://github.com/Emoun/dmutil/blob/ac203f595ad2d87a520c5e5e84fd050d0159a9bb/src/eager_macro_rules.rs#L205-L233)

It has some very weird macros:

```rust
	(	@final[
			$(#[$($metas:tt)*])*
			$macro_name:ident $dollar1:tt $id_1:ident
			$({$($rules_grammar:tt)*} => {$($rules_expansion:tt)*})+
		]
	)=>{
		$(#[$($metas)*])*
		macro_rules! $macro_name{
			$(
				// First the eager supporting version
				{
					@eager[$dollar1($dollar1 $id_1:tt)*]
					$($rules_grammar)*
				} => {
					eager_internal!{
						@from_macro[$dollar1($dollar1 $id_1)*]
						$($rules_expansion)*
					}
				};
			)+
			
			$(
				// Then the pure version. We put the pure versions
				// last such that if it contains a '$($all:tt)*' rule,
				// the pure version will not catch an eager call.
				{$($rules_grammar)*} => {$($rules_expansion)*};
			)+
		}
	};
```


As far as I can tell, `dmutil` isn't doing anything wrong. However, we seem to be choking on the `None`-delimited groups when parsing the deserialized `macro_rules!` body (which is why the error only ocurs when running tests). That is, the parser can handle `NtExpr(<inner_tokens>)` or `<inner_tokens>`, but not `NoneGroup(<inner_tokens>)`.

**Comment by Aaron1011** [comment]
Specifically, we're running into problems here:

https://github.com/rust-lang/rust/blob/c977b8775dd72d191ff1d8e8dceaf4b4cd5db86c/src/librustc_expand/mbe/quoted.rs#L112-L118

We're encountering a `None`-delimited group, but this code is only set up to handle groups with normal delimiters.

@petrochenkov: Should I add in enough `None`-delimited group handling to make `dmutil` compile, or should this PR be blocked on more robust handling of `None`-delimited groups?

File: src/test/ui/proc-macro/auxiliary/meta-delim.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+macro_rules! produce_it {
+    ($dollar_one:tt $foo:ident $my_name:ident) => {
+        #[macro_export]
+        macro_rules! meta_delim {
+            ($dollar_one ($dollar_one $my_name:ident)*) => {
```
Comment (Aaron1011) [inline src/test/ui/proc-macro/auxiliary/meta-delim.rs:5]: :scream:

**Comment by Aaron1011** [comment]
@petrochenkov: I've added a workaround to get `dmutil` compiling. This PR should be ready to merge.

**Comment by petrochenkov** [comment]
@Aaron1011 
Do I understand correctly that in the regression case `NtTT` (`$dollar_one`) is encountered during AST lowering and it gets wrapped into a `None` group?

In that case it's not correct, `NtTT` must not be wrapped into a group, it must be passed as is.
I just assumed that `NtTT` can never be passed to `nt_to_tokenstream`, neither in proc_macro_server, nor in AST lowering, just because it's substituted early.
But apparently that's not the case for generated `macro_rules`.

**Comment by Aaron1011** [comment]
@petrochenkov: The `NtTT` doesn't get wrapped - whatever occurs after the dollar sign (an ident or delimited group) does.

**Comment by petrochenkov** [comment]
Ah, ok, so we get `$ ⟪var⟫` or `$⟪(tokens)⟫*` and `macro_rules` can't parse it because it doesn't do the same automatic flattening of `NoDelim` groups as `TokenCursor::next` (IIRC) does, and sees the original token stream.

**Comment by petrochenkov** [comment]
Seems ok to patch `macro_rules` like this, it's basically the same thing as the attribute changes from the earlier commits, attributes see the original token stream without `NoDelim` group flattening as well.

@bors r+
(The PR could use some squashing to eliminate fixup commits though.)

**Comment by Aaron1011** [comment]
Squashed into more meaningful commits

@bors r=petrochenkov

**Pull request merged**